Update and extend Adapter Software Engineering (#534)
precice/precice.github.io: pages/docs/couple-your-code/couple-your-code-adapter-software-engineering.md

title: Adapter software engineering
permalink: couple-your-code-adapter-software-engineering.html
keywords: api, adapter, library, modularity, sustainability
summary: "The example developed in the step-by-step guide is a rather intrusive way of writing an adapter as we directly modify the main solver routines. This page discusses alternative architectures, as well as further technical aspects of the adapter as a software project."
---

## Architecture

What we develop in the [step-by-step guide](couple-your-code-preparing-your-solver) is best described as an _adapted code_, not an _adapter_. We directly modified the main solver routines. There are also ways to separate the adapter code from the solver code, making the adapter easier to maintain and distribute/publish:

![Adapter software engineering options](images/docs/couple-your-code/adapter-software-engineering/adapter-software-engineering-approaches.svg)

{% tip %}

> In any case, fully integrating the adapter into the codebase and the continuous integration system of the solver (with any of the methods described here) would make maintenance easier, as most complications typically come from changes in the solver code, not from preCICE (in which case, you can refer to a [porting guide](./couple-your-code-porting-overview.html)).

Reviewer: ```suggestion
In any case, fully integrating the adapter into the codebase and the continuous integration system of the solver (with any of the methods described here) makes maintenance easier, as most complications typically come from changes in the solver code, not from preCICE (in which case, you can refer to a [porting guide](./couple-your-code-porting-overview.html)).
```

{% endtip %}

### Adapter class

Encapsulating the adapter into a class separates the preCICE calls from the rest of the solver code.
This is very similar to directly modifying the solver code (with some modifications still needed),
but the solver calls a few high-level adapter functions, instead of directly calling the preCICE API.
Such encapsulation avoids duplicating additional checks and details and makes maintenance easier.

This method is particularly useful for adapters used in frameworks that the user uses to create concrete simulations.
Examples here include the [FEniCS adapter](./adapter-fenics.html) and the [deal.II adapter](https://github.com/precice/dealii-adapter).

Even when directly modifying the code, it makes sense to hide as many changes as possible behind a class API.
Examples for this include the [CalculiX adapter](./adapter-calculix-overview.html) and the [SU2 adapter for v6 and earlier](./adapter-su2-overview.html) (as described in [Alexander Rusch's thesis](https://www5.in.tum.de/pub/Rusch2016_BA.pdf)).

> Adapter classes can still be provided as optional modules of the main codebase. Examples include the [deal.II adapter](./adapter-dealii-overview.html), the [G+smo adapter](./adapter-gismo.html), the [DUNE adapter](./adapter-dune.html), and the [DuMux adapter](./adapter-dumux-get.html)

Reviewer: Would it be easier to understand if we bring the packaging aspect earlier in this section? "Separate class", but then one still has different option on how to maintain and package, "integrated" or "independent"?

Author: There was a hint in the first paragraph, expanded that further in https://github.com/precice/precice.github.io/pull/534/commits/ad8e2f9ba4f703e0b06287dd4941abf904159f2e

### Adapter plugin

Some solvers provide a plugin interface.
Examples include OpenFOAM (function objects) and Fluent (UDF files).
The solver exposes specific "hooks"/"callbacks", which trigger the execution of external code at specific stages of the simulation.
Creating an adapter as such a plugin is non-intrusive and makes it easy to distribute the adapter independently.

While this is a generally preferable architecture, it comes with some requirements.
The solver needs to provide the following functionality via the plugin interface:

- Executing additional steps during the configuration, setup, and teardown
- Access to the (boundary) mesh coordinates (read)
- Access to the (boundary) values (read/write)
- Access to the time step size (read/write)
- Controlling the exit condition of the time loop

> - Checkpointing:

Reviewer: ```suggestion
- Checkpointing (if implicit coupling is required):
```
Not sure if here is the best place, but we should mention this.

  - Storing and restoring the time
  - Storing and restoring the internal solution

The trickiest part is typically the checkpointing. However, in some cases, this can alternatively be implemented as (a) controlling when the solver moves to the next time step or discards the current solution and tries again, (b) storing restart files and restarting from them (inefficient, but might not be an issue in some cases).

Examples of adapters implemented as plugins include the [OpenFOAM adapter](./adapter-openfoam-overview.html) (with the plugin approach explained in detail in [Gerasimos Chourdakis' thesis](https://www5.in.tum.de/pub/Chourdakis2017_Thesis.pdf)) and the [Fluent adapter](https://github.com/precice/fluent-adapter).

### Adapter wrapper

Some solvers provide an API, meaning that specific functions of the solver can be executed from external programs.
A common example is a Python API, which allows us to write a Python wrapper for the solver.

> This wrapper then can call both the solver and preCICE, making the wrapper an adapter.

Reviewer: We are in some gray area here: Also FEniCS has an API, as an example. Is an FEniCS application script then also an adapter? Not sure how to best solve this, but maybe we could write somewhere that these categories are ambiguous to some extent.

Author: Added a disclaimer on the unclarity of the categorization in the introduction via https://github.com/precice/precice.github.io/pull/534/commits/3917e462af5cf41b848cc0862872b71ba8167211.

I think the wrapper category is still distinct, and the section helps developers see this possibility.

{% endtip %}

### Configuration format

An adapter configuration file can be written in any configuration format.
Choose a format that is easy for the user of the respective solver to understand and write,
as well as easy for the adapter to parse.
Common standard formats include [YAML](https://en.wikipedia.org/wiki/YAML) and [JSON](https://en.wikipedia.org/wiki/JSON), for which we develop generation and validation tools.
Both are used by existing adapters and have good tooling support.
YAML is primarily designed for configuration files, it is more convenient to read and write, and supports comments.
JSON is primarily designed for data interchange, focuses on interoperability, and is often the primary format that some tools use.
In the future, we aim to have the adapter configuration files automatically generated by default.

### Adapter configuration schema

We provide a standard [preCICE adapter configuration schema](https://github.com/precice/preeco-orga/tree/main/adapter-config-schema) to facilitate the interoperability between adapters.
Expecting configuration in this general format has several benefits:

> - allows you to generate and validate adapter configuration files using the provided preCICE tools (including the [MetaConfigurator](https://www.metaconfigurator.org/))

Reviewer 2: ```suggestion
- allows you to generate and validate adapter configuration files using the provided preCICE tools (including the [MetaConfigurator](https://www.metaconfigurator.org/?schema=https://github.com/precice/preeco-orga/blob/main/adapter-config-schema/precice_adapter_config_schema.json))
```

Reviewer 2: Optional: this could be done for user convenience, so they will not only see MetaConfigurator but also have the adapter schema pre-loaded. Notice, that we also have a custom MetaConfigurator settings file, which I have not pre-loaded here, because it disables the Schema Editor view and other things and this link is intended to show MetaConfigurator, rather than start an adapter editing session. The same argument could be made in favor of not pre-loading the schema, but I think with the schema it will become more evident to the user why such a tool is useful and what it can do.

Author: Very good suggestions, thanks!
I start noticing, however, that one needs to keep track of these specific URLs when starting. What is the plan for making this schema easier to find?

Publish in the JSON schema store?

Add a drop-down for common use cases? (preCICE, chemistry application)

Reviewer 2: These URLs provide a very reliable and robust ways to pre-load a schema, which does not need any third party service or data base other than MetaConfigurator and the source of the schema. It is useful if someone wants to share a link with someone that pre-loads a schema, data or settings. --> creators of schemas/tools can share a link to MetaConfigurator with their schema

A more easy way for users to share their schema and data is by using the internal share button in MetaConfigurator, which stores the schema, data and settings at a university server and then provides a unique URL that opens MetaConfigurator with these files. --> anyone can generate and share a link with their schema and data with one click

Now, for someone who opens MetaConfigurator by themselves without any external link and wants to find a certain schema, the JSON Schema store is a central place to find most schemas. We should also upload the adapter schema there at one point. 

For the future, we have planned some more integration of existing (academic) data storage systems, such as DaRUS, into MetaConfigurator. With these systems, MetaConfigurator would offer more options of finding schemas or data and also for submitting/persisting a schema or data.

Edit: also, this URLs are just a way that makes it more convenient to share schemas. Of course, the user can also just press the "Open schema" button and open the schema from their local file system or an URL. Or they can just copy and paste the schema into the editor or drag and drop it into it.

Author: Alright, then let's point to the concrete URL for now. It actually fits nicely in the sentence: "here is a schema, and here is a MetaConfigurator URL using that schema".
